[scope] [coq] Make Coq_lib.DB creation lazy

ejgallego opened the pull request:
This will be important once we have installed_theories in #7047

ejgallego committed:
[scope] [coq] Make Coq_lib.DB creation lazy

This will be important once we have installed_theories in #7047

Signed-off-by: Emilio Jesus Gallego Arias <e+git@x80.org>

Alizter commented on src/dune_rules/scope.ml:
```diff
@@ -154,13 +154,15 @@ module DB = struct
         else None)
     |> Coq_lib.DB.create_from_coqlib_stanzas ~find_db ~parent:None
 
+  let public_theories ~find_db coq_stanzas = Lazy.from_fun (public_theories ~find_db coq_stanzas)
```
Isn't it more efficient to just use `lazy` above?

ejgallego replied:
No

Alizter commented on src/dune_rules/scope.ml:
```diff
@@ -265,14 +268,16 @@ module DB = struct
       |> Path.Source.Map.map ~f:(fun stanzas ->
-             Coq_lib.DB.create_from_coqlib_stanzas
-               ~parent:(Some public_theories) ~find_db stanzas)
+             Lazy.from_fun (fun () ->
```
This can just be `lazy` I guess

Alizter commented on CHANGES.md:
```diff
@@ -3,6 +3,7 @@ Unreleased
 
 - Fix preludes not being recorded as dependencies in the `(mdx)` stanza (#7109,
   fixes #7077, @emillon).
+- Make coq_db creation in scope lazy (@ejgallego, #7133)
```
I am not sure this is going to be noticeable for users.

ejgallego replied:
Hopefully it is not.

ejgallego replied:
To clarify why I added the changelog, as this is a large change in (internal) behavior, Dune Coq users appreciate having a list of things to look at if their setup breaks at upgrade time.

Alizter commented:
Hi @rgrinberg we are touching scope.ml, can you take a look?

rgrinberg commented on src/dune_rules/scope.ml:
```diff
@@ -265,14 +268,16 @@ module DB = struct
                   match stanza.package with
                   | None -> Coq_lib.DB.Theory dir
-                  | Some _ -> Redirect public_theories
+                  | Some _ -> Redirect (Lazy.force public_theories)
```
Might as well go the whole way and make coq_db_by_project_dir map lazily constructed.

ejgallego replied:
How's that gonna work for merging the map with `db_by_project_dir` ? We still need access to the keys.

rgrinberg replied:
Instead of merging, you can just look it up after forcing the map:

```ocaml
let coq_db = lazy (Path.Source.Map.find_exn (Lazy.force coq_map) dir)) in ...
```

ejgallego replied:
Ok, that's a bit deeper change, let's see how it looks like.

ejgallego replied:
Done

ejgallego replied:
Turned out great in the end IMHO.

ejgallego replied:
Ok I found the cause, the problem is that now library DB creation is too lazy, and the install rules run first emitting the more confusing error:
```  
$ dune build B
 |  Error: Cannot have more than one boot theory in scope:
 |  - Coq at Coq/dune:1
 |  - Coq at B/Coq/dune:2
+|  -> required by alias B/default
+|  Error: Multiple rules generated for
+|  _build/install/default/lib/coq/theories/Init/Prelude.v:
+|  - Coq/dune:1
+|  - B/Coq/dune:2
+|  -> required by _build/default/B/Foo.install
+|  -> required by alias B/all
+|  -> required by alias B/default
```
I'm tempted to leave this like they are in this PR, even with the worse error message. WDTY @rgrinberg ?

rgrinberg replied:
Fine by me. You can also easily fix bring back the old behavior if you force the coq scope when setting up install rules for a theory.

ejgallego replied:
Indeed the latter option seems the best for me, thanks.

ejgallego committed:
[scope] [coq] Make the full Coq scope creation lazy

Note the hack in coq_rules as to have better errors in case
of conflicting theories.

Signed-off-by: Emilio Jesus Gallego Arias <e+git@x80.org>

ejgallego committed:
[scope] [coq] Refactoring towards a Coq scope

Signed-off-by: Emilio Jesus Gallego Arias <e+git@x80.org>

ejgallego commented:
I did stop short of putting Coq scopes in their own type as Ali suggested because we would have to make the DB struct generic, and then quite a few changes in the rules, but maybe indeed it is worth it.

ejgallego commented:
Ok, I've restored the more sensible error message, this is ready IMO.

Note the messages are a bit different but IMHO that's a good thing?

ejgallego committed:
Merge branch 'main' into lazy_coq_Db
